pkg/operator/controller/externaldns/status.go

```diff
@@ -0,0 +1,296 @@
+	if err != nil {
+		return metav1.Condition{
+			Type:    ExternalDNSDeploymentReplicasMinAvailableConditionType,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "InvalidMaxUnavailableValue",
+			Message: fmt.Sprintf("invalid value for max unavailable: %v", err),
+		}
+	}
+	if maxSurge == 0 && maxUnavailable == 0 {
+		//Use a default value here
+		maxUnavailable = 1
+	}
+	if int(deployment.Status.AvailableReplicas) < int(replicas)-maxUnavailable {
+		return metav1.Condition{
+			Type:    ExternalDNSDeploymentReplicasMinAvailableConditionType,
+			Status:  metav1.ConditionFalse,
+			Reason:  "DeploymentMinimumReplicasNotMet",
+			Message: fmt.Sprintf("%d/%d of replicas are available, max unavailable is %d", deployment.Status.AvailableReplicas, replicas, maxUnavailable),
+		}
+	}
+
+	return metav1.Condition{
+		Type:    ExternalDNSDeploymentReplicasMinAvailableConditionType,
+		Status:  metav1.ConditionTrue,
+		Reason:  "DeploymentMinimumReplicasMet",
+		Message: "Minimum replicas requirement is met",
+	}
+}
+
+// computeAllReplicasCondition returns an externalDNS condition based on the deployment, its number of desired pods
+// and the number of available replicas.
+// the condition is true if the number of available replicas is greater than number of desired replicas
+func computeAllReplicasCondition(deployment *appsv1.Deployment) metav1.Condition {
+	replicas := int32(1)
+	if deployment.Spec.Replicas != nil {
+		replicas = *deployment.Spec.Replicas
+	}
+
+	if deployment.Status.AvailableReplicas < replicas {
+		return metav1.Condition{
+			Type:    ExternalDNSDeploymentReplicasAllAvailableConditionType,
+			Status:  metav1.ConditionFalse,
+			Reason:  "DeploymentReplicasNotAvailable",
+			Message: fmt.Sprintf("%d/%d of replicas are available", deployment.Status.AvailableReplicas, replicas),
+		}
+	}
+
+	return metav1.Condition{
+		Type:    ExternalDNSDeploymentReplicasAllAvailableConditionType,
+		Status:  metav1.ConditionTrue,
+		Reason:  "DeploymentReplicasAvailable",
+		Message: "All replicas are available",
+	}
+}
+
+// computeDeploymentPodsScheduledCondition lists the pods matching the namespace and the label selector of the deployment
+// returns condition true when all matching pods are scheduled
+// returns condition false if no matching pods were found, or if any of the pods are unscheduled
+func computeDeploymentPodsScheduledCondition(ctx context.Context, cl client.Client, deployment *appsv1.Deployment) metav1.Condition {
+	selector, err := metav1.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil || selector.Empty() {
+		return createPodsScheduledUnknownCondition("InvalidLabelSelector", "Deployment has an invalid label selector.")
+	}
+	pods, err := getFilteredPodsList(ctx, cl, deployment.Namespace, selector)
+	if err != nil {
+		return createPodsScheduledUnknownCondition("PodScheduledUnknown", "unable to list pods")
+	} else {
```

Comment:
More go idiomatic way of doing this would be to remove `else` as previous `if` would return if there is an error.

Reply:
ok